src/group_theory/divisible.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import group_theory.subgroup.pointwise
+import group_theory.quotient_group
+import data.real.basic
+import data.complex.basic
+
+/-!
+# Divisible Group
+
+An abelian group `A` is said to be divisible iff for all `n ≠ 0` `y ∈ A`, there is an `x ∈ A`
+such that `n • x = y`.
+
+## Main results
+
+* `add_comm_group.divisible_iff_smul_top_eq_top` : `A` is divisible iff `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.divisble_rat` : `ℚ` is a divisible group.
+* `add_comm_group.divisible_real` : `ℝ` is a divisible group.
+* `add_comm_group.divisible_complex` : `ℂ` is a divisible group.
+* `add_comm_group.divisible_pi` : Any product of divisble group is divisible.
+* `add_comm_group.divisible_quotient : Quotient group of divisible group is divisible.
+* `add_comm_group.divisible_of_surj` : if `A` is divisible and `A →+ B` is surjective, then `B` is
+  divisible.
+TODO: Show that divisibility implies injectivity in the category of `AddCommGroup`.
+-/
+
+
+namespace add_comm_group
+
+open_locale pointwise
+
+variables (A : Type*) [add_comm_group A]
+
+/--
+A divisible group `A` is an abelian group such that `nA = A` for all `n ≠ 0`.
+-/
+class divisible : Prop :=
+(div : ∀ (n : ℤ), n ≠ 0 → function.surjective ((•) n : A → A))
+
+
+lemma divisible_iff_smul_top_eq_top :
+  divisible A ↔ ∀ (n : ℤ), n ≠ 0 → n • (⊤ : add_subgroup A) = ⊤ :=
+{ mp := λ ⟨d⟩ n hn,  (add_subgroup.eq_top_iff' (n • ⊤)).mpr $
```

Comment:
I suspect you can get there a lot faster by unfolding `n • (⊤ : add_subgroup A) = ⊤` to ` ⊤.map _ = ⊤` then using `⊤.map _ = _.range`